keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,556 @@
+
+> TODO: More details on the implementation will be charted once the summary, motivation, goals, non-goal and proposal are approved.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_step_invocations_total`
+
+This will be incremented on round-trip of an authorizer. It will track total authorization decision invocations across the following labels.
+
+- `mode` {RBAC, Node, Webhook}
+- `decision` {Allow, Deny, NoOpinion}
+
+2. `apiserver_authorization_step_webhook_invocations_total`
+
+This will be incremented on round-trip of an authorization webhook. It will track total invocation counts across the following labels.
+
+- `server`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+3. `apiserver_authorization_step_webhook_duration_seconds`
+
+This metric will track the average latency
+
+Labels {along with possible values}:
+- `server`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+4. `apiserver_authorization_step_webhook_error_total`
+
+This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.
+
+Labels {along with possible values}:
+
+- server
+- `code` {4xx, 5xx}
+- `decision` {Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Unit tests would be added along with any new code introduced.
+
+Existing test coverage:
+- cmd/kube-apiserver/app/options: 65
+- staging/src/k8s.io/apiserver/pkg/authorization: NA
+- staging/src/k8s.io/apiserver/apis/config/v1: 54.5
+
+> Will add more packages to the list during implementation.
+
+##### Integration tests
+
+Integration tests would be added to ensure the following:
+- Authorization of requests work in the existing flag based mode (feature flag turned off)
+- Authorization of requests work with an apiserver bootstrapped with authorization configuration file (feature flag turned on)
+    - without an webhook
+    - with an webhook - successful request
+    - with an webhook - error on request with `onError: Deny`
+    - with an webhook - error on request with `onError: NoOpinion`
+
+##### e2e tests
+
+End-to-end tests won't be needed as unit and integration tests would cover all the scenarios.
+
+### Graduation Criteria
+
+#### Alpha (1.25)
+
+- Add file based authorizer chain configuration
+- Add feature flag for gating usage
+- Unit tests and Integration tests to be written
+
+#### Beta
+
+- Address user reviews and iterate (if needed keep in Alpha until changes stabilize)
+- Feature flag will be turned on by default
+
+#### GA
+
+- Feature flag removed
+- Existing command line flags will have no effect
```

Comment:
I do not think the existing CLI flag can start deprecation until this new stuff is GA.

Reply:
You would also need to deprecate ABAC since the new config does not support it.